src/main/java/cn/nukkit/Player.java

```diff
@@ -658,91 +795,174 @@ public boolean isPlayer() {
         return true;
     }
 
+    /**
+     * Remove achievement from player if the player has it
+     * @param achievementId achievement id
+     */
     public void removeAchievement(String achievementId) {
         achievements.remove(achievementId);
     }
 
+    /**
+     * Check whether player has an achievement
+     * @param achievementId achievement id
+     * @return has achievement
+     */
     public boolean hasAchievement(String achievementId) {
         return achievements.contains(achievementId);
     }
 
+    /**
+     * Check whether player is still connected
+     * @return connected
+     */
     public boolean isConnected() {
         return connected;
     }
 
+    /**
+     * Get player's display name. Default value is player's username.
+     * @return display name
+     */
     public String getDisplayName() {
         return this.displayName;
     }
 
+    /**
+     * Set player's display name
+     * @param displayName display name
+     */
     public void setDisplayName(String displayName) {
+        if (displayName == null) {
+            displayName = "";
+            server.getLogger().debug("Warning: setDisplayName: argument is null", new Throwable());
+        }
         this.displayName = displayName;
         if (this.spawned) {
-            this.server.updatePlayerListData(this.getUniqueId(), this.getId(), this.getDisplayName(), this.getSkin(), this.getLoginChainData().getXUID());
+            this.server.updatePlayerListData(this.getUniqueId(), this.getId(), this.displayName, this.getSkin(), this.loginChainData.getXUID());
         }
     }
 
     @Override
     public void setSkin(Skin skin) {
         super.setSkin(skin);
         if (this.spawned) {
-            this.server.updatePlayerListData(this.getUniqueId(), this.getId(), this.getDisplayName(), skin, this.getLoginChainData().getXUID());
+            this.server.updatePlayerListData(this.getUniqueId(), this.getId(), this.displayName, skin, this.loginChainData.getXUID());
         }
     }
 
+    /**
+     * Get player's host address
+     * @return host address
+     */
     public String getAddress() {
         return this.socketAddress.getAddress().getHostAddress();
     }
 
+    /**
+     * Get the port of player's connection
+     * @return port
+     */
     public int getPort() {
         return this.socketAddress.getPort();
     }
 
+    /**
+     * Get player's socket address
+     * @return socket address
+     */
     public InetSocketAddress getSocketAddress() {
         return this.socketAddress;
     }
 
+    /**
+     * Get most recent position of received movements
+     * @return next position or current position if no next position has been received
+     */
     public Position getNextPosition() {
         return this.newPosition != null ? new Position(this.newPosition.x, this.newPosition.y, this.newPosition.z, this.level) : this.getPosition();
     }
 
+    private static final Vector3 ZERO_VECTOR3 = new Vector3(0, 0, 0);
```

Comment:
Why is this here? It should be in Vector3 exposed publicly.

Reply:
Because we don't want anyone to access it and accidentally modify its xyz

Reply:
True, the method wasn't meant to be used other than internally